helm: add helm tutorial (#12)
nodeshift/tutorial: helm/README.md

# Deploying your Node.js App via Helm

### Packaging up your Node.js application into a Helm Chart

This will show you how to create your own Helm Chart for your Node.js Application and how to tweak it to best fit your applications needs.

In this self-paced tutorial you will:

- Create a Helm Chart
- Add a dependency chart
- Create liveiness probes

> The application you will use is the one created in the previous tutorial - https://github.com/nodeshift/mern-workshop

Reviewer: ```suggestion
The application you will use is the one created in the previous tutorial - https://github.com/nodeshift/mern-workshop
```

### Prerequisites

> Before getting started, make sure you have the following prerequisites installed on your system.

Reviewer: Maybe I already asked this, but would it be possible to use the sandbox instead of a local install?

Author: It probably can but I thought we werent going to promote the sandbox in this tutorial.

Reviewer: I guess I just come back to it knowing how tricky/time consuming setting up kube can be, especially if people don't want to add to the config on their machine.  That might not help completely in this case though since they'd still need docker and helm installed locally.

```sh
sudo apt-get install docker-ce docker-ce-cli containerd.io
```

More information can be found at https://docs.docker.com/install/linux/docker-ce/ubuntu/

Add the user to the docker group (optional for part 1, required for part 2 and 3)

```sh
sudo groupadd docker
sudo gpasswd -a $USER docker
sudo service docker restart
```

Install docker compose

```sh
sudo apt install docker-compose
```

</details>

#### `microk8s`

<details>

```sh
snap install --channel 1.14/stable microk8s --classic
sudo usermod -a -G microk8s ibmuser
sudo microk8s.start
sudo snap alias microk8s.kubectl kubectl
export PATH=/snap/bin:$PATH
sudo microk8s.config >~/.kube/config
microk8s.enable dns registry
```

You may be prompted to add your userid to the 'microk8s' group to avoid having to use `sudo` for all the commands.

**Note**: The Prometheus Helm chart is
[not compatible](https://github.com/helm/charts/pull/17268) with
Kubernetes 1.16, so make sure to install 1.14.

</details>

#### `minikube`

<details>

```sh
minikube start --kubernetes-version=1.14.7
eval $(minikube docker-env)
```

**Note** that the Prometheus Helm chart is
[not compatible](https://github.com/helm/charts/pull/17268) with

> Kubernetes 1.16, so make sure to run with 1.14.7.

Reviewer: I don't think this part is right since it later requires a newer version of kube

```

These files will form the basis of your Helm Chart, lets explore the precreated files and make some nessacary changes

### 2. Editing the .helmignore file

   This file works the same as any .ignore file, you just fill in the patterns you don't want to be packaged up into the helm chart. For example if you have some secrets saved as a JSON file that you dont want to be inside the helm chart. For our application we dont have any files we need to protect so lets move on to the next step.

### 3. The Chart.yaml file

   This file contains the base information about your Helm Chart, your premade one should look similar to this (with extra comments):
   
   ```yaml
   apiVersion: v2
   name: myapp
   description: A Helm chart for Kubernetes
   type: application
   version: 0.1.0
   appVersion: 1.0.0
   ```

   `apiVersion: v2` signals that this chart is designed for Helm3 support _only_.

   `version` is the charts version, increment this under semver everytime you update the chart

   `appVersion` is the version of the app you are deploying, this is to be increased everytime you increase the version of your app but does not impact the charts version 

>    The rest of the fields are self explanatory but lets add some more information to describe our chart. We are going to set a Kubernetes minimum version, add some descriptive keywords and add our name as Maintainers. So go ahead and add the following to your `Chart.yaml` whilst subsituting your name and email in:

Reviewer: Does it matter where the additional yml is added? If it needs to be at the end instead of the front that might make sense to say.

Reviewer 2: ```suggestion
   The rest of the fields are self-explanatory but let's add some more information to describe our chart. We are going to set a Kubernetes minimum version, add some descriptive keywords and add our name as Maintainers. So go ahead and add the following to your `Chart.yaml` whilst substituting your name and email in:
```

   ```yaml
   kubeVersion: '>= 1.21.0-0'
   keywords:   
   - nodejs
   - express
   - mern
   maintainers:
   - name: Firstname Lastname
     email: FirstnameLastname@company.com
   ```

   The final key thing we are going to add is a dependecy, our application needs mongoDB to run so we are going to call a premade mongo chart to install mongo as we install our chart. Firstly we need to add to our `Chart.yaml`:

   ```yaml
   dependencies:
   - name: mongodb

>      version: 5.0.2

Reviewer: This did not work for me. I had to upgrade kube to pass the kubeVersion line added above. After a while I believe I then figured out that for 5.0.2 helm failed the install with the complaint:

```
Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: unable to recognize "": no matches for kind "Deployment" in version "extensions/v1beta1"
```

```

Reviewer: I've updated to 10.0.0 for the mondodb version, and it installs, I can hit the app but it does not add items to the todo list

Reviewer: Seems like some of the options changed, added comment to update values for mongodb and its working now

>      version: 5.0.2

Reviewer: ```suggestion
     version: 10.0.0
```

Reviewer: Note it also seems to work with 8.0.0 if 10.0.0 is too recent for some reason on other platforms.

>     version: 5.0.2
>     repository: https://charts.bitnami.com/bitnami

Reviewer: ```suggestion
     version: 5.0.2
     repository: https://charts.bitnami.com/bitnami
```

   ```

   Then run the following command in terminal to download the chart:
   
   ```sh
   helm dependency update
   ```

### 4. Template files

Inside the templates folder you will find that Helm has created some files for us already:

```sh
$ ls myapp/templates

NOTES.txt
_helpers.tpl
deployment.yaml
hpa.yaml
ingress.yaml
service.yaml
serviceaccount.yaml
tests
```

These files represent the kubernetes objects that our Helm Chart will deploy - you can set up the deployment and service, and you can also create a serviceaccount for your app to use. The `NOTES.txt` file is just the notes that will be displayed to the user when they deploy your Helm Chart, you can use this to provide further instruction to them to get your application working. `_helpers.tpl` is where template helpers that you can re-use throughout the chart go. However for this tutorial we are going to use our own files so you can go ahead and remove the precreated files:

```sh

> $ rm myapp/templates/*

Reviewer: ```suggestion
$ rm -rf myapp/templates/*
```

```

Now lets move on to making our own template files!

#### 4.1 Frontend Service

The first template we will create will deploy the service for our frontend. Lets create a file called `chart/myapp/templates/frontend-service.yaml` and paste in the following:

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/scrape: 'true'
  name: "frontend-service"
spec:
  ports:
  - name: http
    port: {{ .Values.frontend.service.servicePort }}
    nodePort: 30444
  type: NodePort
  selector:
    app: "frontend-selector"
```

> What this create will the service that serves our frontend and allows for it to be connected to the outside world. `nodePort` is the port we want our service to be accessible through so `localhost:30444` will take us to our frontend once deployed

Reviewer: ```suggestion
This will create the service that serves our frontend and allows for it to be connected to the outside world. `nodePort` is the port we want our service to be accessible through so `localhost:30444` will take us to our frontend once deployed
```

apiVersion: apps/v1
kind: Deployment
metadata:
  name: frontend-deployment
  labels:
    chart: frontend
spec:
  selector:
    matchLabels:
      app: "frontend-selector"
  template:
    metadata:
      labels:
        app: "frontend-selector"
    spec:
      containers:
      - name: frontend
        image: "{{ .Values.frontend.image.repository }}:{{ .Values.frontend.image.tag }}"
        imagePullPolicy: {{ .Values.frontend.image.pullPolicy }}
        livenessProbe:
          httpGet:
            path: /
            port: {{ .Values.frontend.service.servicePort }}
          initialDelaySeconds: {{ .Values.frontend.livenessProbe.initialDelaySeconds}}
          periodSeconds: {{ .Values.frontend.livenessProbe.periodSeconds}}
        ports:
        - containerPort: {{ .Values.frontend.service.servicePort}}
```

What this creates is our frontend deployment, inside that is the pod that is running our frontend image plus its replicaset.

`image` is the image we want the container to deploy using variables pulled from our `values.yaml` file.

We also add the `frontend-selector` label to our pod which allows for our frontend service to connect with it.

#### 4.3 Backend Service 

> Our third file will spawn the service for the backend part of our application. Create a file called `chart/myapp/templates/backend-service.yaml` and paste in the following:

Reviewer: Maybe we should remove `chart` from the path here as it was not in the previous line removing the templates

        app: "backend-selector"
    spec:
      containers:
      - name: backend
        image: "{{ .Values.backend.image.repository }}:{{ .Values.backend.image.tag }}"
        imagePullPolicy: {{ .Values.backend.image.pullPolicy }}
        livenessProbe:
          httpGet:
            path: /health
            port: {{ .Values.backend.service.servicePort }}
          initialDelaySeconds: {{ .Values.backend.livenessProbe.initialDelaySeconds}}
          periodSeconds: {{ .Values.backend.livenessProbe.periodSeconds}}
        ports:
        - containerPort: {{ .Values.backend.service.servicePort}}
        env:
          - name: PORT
            value : "{{ .Values.backend.service.servicePort }}"
          - name: APPLICATION_NAME
            value: "{{ .Release.Name }}"
          - name: MONGO_URL
            value: {{ .Values.backend.services.mongo.url }}
          - name: MONGO_DB_NAME
            value: {{ .Values.backend.services.mongo.name }}
```

Similar to our frontend deployment file, this file creates our backend deployment, with a key difference being the mongoDB information that is passed through to allow for communication with the mongo instance.

### 5. Values file

> For the `values.yaml` file we are going to split it in 3 sections, have a read of each section and then add them all to your `values.yaml` file

Reviewer: Since we specified the path for the other files we should probably include it here as well.

```yaml
# Frontend
frontend:
  replicaCount: 1
  revisionHistoryLimit: 1
  image:
    repository: frontend 
    tag: v1.0.0
    pullPolicy: IfNotPresent
    resources:
      requests:
        cpu: 200m
        memory: 300Mi
  livenessProbe:
    initialDelaySeconds: 30
    periodSeconds: 10
  service:
    name: frontend
    type: NodePort
    servicePort: 80 # the port where nginx serves its traffic
```
These values are for the frontend section. Here we pass through the image name, tag and the values for the frontend service.

```yaml
# backend
backend:
  replicaCount: 1
  revisionHistoryLimit: 1
  image:
    repository: backend
    tag: v1.0.0
    pullPolicy: IfNotPresent
    resources:
      requests:
        cpu: 200m
        memory: 300Mi
  livenessProbe:
    initialDelaySeconds: 30
    periodSeconds: 10
  service:
    name: backend
    type: NodePort
    servicePort: 30555
  services:
    mongo:
      url: mongo-mongodb
      name: todos
      env: production
```

These values are for the backend section. Here we pass through the image, tag, service information, and some mongoDB information to locate the instance.

```yaml
# mongo

> mongo:
>   usePassword: false
>   replicaSet.enabled: true
>   service.type: LoadBalancer
>   replicaSet.replicas.secondary: 3

Reviewer: ```suggestion
mongodb:
  auth:
    enabled: false
  replicaSet.enabled: true
  service.type: LoadBalancer
  replicaSet.replicas.secondary: 3
```

Reviewer: I think the parameters were not taking effect as `mongo` needs to match `mongodb` as that is the name of the dependency in the Chart.yaml file.

In addition for newer version of mongodb usePassword was replaced with the auth.enabled as show in the change.

```

Finally these values are passed through to the mongoDB chart we downloaded earlier.

### 6. Deploy your Helm Chart

First we need to build our docker images to deploy, from the root directory of the project run:

```sh
$ docker build -f frontend/Dockerfile -t frontend:v1.0.0 frontend

> $ docker build -f backend/Dockerfile -t backend:v1.0.0 backend

Reviewer: @AshCripps if you've already updated, I think its still missing the instructions for cloning  https://github.com/nodeshift/mern-workshop and being in the right place to run the comments.

```

Once the images are built you can now deploy your helm chart

```sh
$ helm install myapp chart/myapp
```

### Congratulations! 🎉

> You have now created a Helm Chart which deploys a frontend and a backend whilst also calling and installing a dependency chart from the internet.

Reviewer: Maybe add a link here to the doc in the mern tutorial that shows how to test out the running app and possibly a few paragraphs running kubectl commands to look at the deployed pods?

Reviewer 2: Agreed, we need the `kubectl get pods` steps to demonstrate the three pods the Helm chart has created. And then it'd be good to instruct the user how to access the frontend and add an item to the todo list that is persisted in MongoDB (just to prove the end-to-end flow).

# Deploying your Node.js App via Helm

### Packaging up your Node.js application into a Helm Chart

This will show you how to create your own Helm Chart for your Node.js Application and how to tweak it to best fit your applications needs.

In this self-paced tutorial you will:

- Create a Helm Chart
- Add a dependency chart

> - Create liveiness probes

Reviewer 2: ```suggestion
- Create liveness probes
```

</details>

#### `minikube`

<details>

```sh
minikube start --kubernetes-version=1.14.7
eval $(minikube docker-env)
```

</details>

#### Installing Helm

Helm is a package manager for Kubernetes. By installing a Helm "chart" into your Kubernetes cluster you can quickly run all kinds of different applications. You can install Helm using one of the options below:

**Using a Package Manager:**

* macOS with Homebrew: `brew install helm`
* Linux with Snap: `sudo snap install helm --classic`
* Windows with Chocolatey: `choco install kubernetes-helm`

**Using a Script:**

```sh
$ curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3
$ chmod 700 get_helm.sh
$ ./get_helm.sh
```

#### Downloading the application

Make sure you clone down the application repo and you are in the correct directory.

```sh
$ git clone https://github.com/nodeshift/mern-workshop.git
$ cd mern-workshop
```

### 1. Create your Helm Chart files

In your terminal inside the folder holding your application run:

```sh
$ mkdir chart
$ cd chart
$ helm create myapp
```

> This will create the following filestructure:

Reviewer 2: ```suggestion
This will create the following file structure:
```

```
myapp/
├── .helmignore   # Contains patterns of which files to ignore when packaging your Helm Chart.
├── Chart.yaml    # Information about your chart
├── values.yaml   # The default values for your templates
├── charts/       # Charts that this chart depends on
└── templates/    # The template files
    └── tests/    # The test files
```

> These files will form the basis of your Helm Chart, lets explore the precreated files and make some nessacary changes

Reviewer 2: ```suggestion
These files will form the basis of your Helm Chart, lets explore the pre-created files and make some necessary changes.
```

### 2. Editing the .helmignore file

>    This file works the same as any .ignore file, you just fill in the patterns you don't want to be packaged up into the helm chart. For example if you have some secrets saved as a JSON file that you dont want to be inside the helm chart. For our application we dont have any files we need to protect so lets move on to the next step.

Reviewer 2: ```suggestion
This file works the same as any .ignore file, you just fill in the patterns you don't want to be packaged up into the helm chart. For example, if you have some secrets saved as a JSON file that you don not want to be inside the helm chart. For our application we dont not have any files we need to protect so let's move on to the next step.
```

### 3. The Chart.yaml file

   This file contains the base information about your Helm Chart, your premade one should look similar to this (with extra comments):
   
   ```yaml
   apiVersion: v2
   name: myapp
   description: A Helm chart for Kubernetes
   type: application
   version: 0.1.0
   appVersion: 1.0.0
   ```

   `apiVersion: v2` signals that this chart is designed for Helm3 support _only_.

>    `version` is the charts version, increment this under semver everytime you update the chart

Reviewer 2: ```suggestion
   `version` is the charts version, increment this under semver every time you update the chart
```

>    `appVersion` is the version of the app you are deploying, this is to be increased everytime you increase the version of your app but does not impact the charts version 

Reviewer 2: ```suggestion
   `appVersion` is the version of the app you are deploying, this is to be increased every time you increase the version of your app but does not impact the charts version 
```

   The rest of the fields are self explanatory but lets add some more information to describe our chart. We are going to set a Kubernetes minimum version, add some descriptive keywords and add our name as Maintainers. So go ahead and add the following to your `Chart.yaml` whilst subsituting your name and email in:

   ```yaml
   kubeVersion: '>= 1.21.0-0'
   keywords:   
   - nodejs
   - express
   - mern
   maintainers:
   - name: Firstname Lastname
     email: FirstnameLastname@company.com
   ```

>    The final key thing we are going to add is a dependecy, our application needs mongoDB to run so we are going to call a premade mongo chart to install mongo as we install our chart. Firstly we need to add to our `Chart.yaml`:

Reviewer 2: ```suggestion
   The final key thing we are going to add is a dependency, our application needs mongoDB to run so we are going to call an existing mongo chart to install mongo as we install our chart. Firstly we need to add to our `Chart.yaml`:
```

   ```yaml
   dependencies:
   - name: mongodb
     version: 10.26.3
     repository: https://charts.bitnami.com/bitnami
   ```

>    Then run the following command in terminal to download the chart:

Reviewer 2: ```suggestion
   Then run the following command in the terminal to download the chart:
```

   
   ```sh
   helm dependency update
   ```

### 4. Template files

Inside the templates folder you will find that Helm has created some files for us already:

```sh
$ ls myapp/templates

NOTES.txt
_helpers.tpl
deployment.yaml
hpa.yaml
ingress.yaml
service.yaml
serviceaccount.yaml
tests
```

> These files represent the kubernetes objects that our Helm Chart will deploy - you can set up the deployment and service, and you can also create a serviceaccount for your app to use. The `NOTES.txt` file is just the notes that will be displayed to the user when they deploy your Helm Chart, you can use this to provide further instruction to them to get your application working. `_helpers.tpl` is where template helpers that you can re-use throughout the chart go. However for this tutorial we are going to use our own files so you can go ahead and remove the precreated files:

Reviewer 2: ```suggestion
These files represent the kubernetes objects that our Helm Chart will deploy - you can set up the deployment and service, and you can also create a serviceaccount for your app to use. The `NOTES.txt` file is just the notes that will be displayed to the user when they deploy your Helm Chart, you can use this to provide further instruction to them to get your application working. `_helpers.tpl` is where template helpers that you can re-use throughout the chart go. However for this tutorial we are going to use our own files so you can go ahead and remove the generated files:
```

```sh
$ rm -rf chart/myapp/templates/*
```

> Now lets move on to making our own template files!

Reviewer 2: ```suggestion
Now let's move on to making our own template files!
```

#### 4.1 Frontend Service

> The first template we will create will deploy the service for our frontend. Lets create a file called `chart/myapp/templates/frontend-service.yaml` and paste in the following:

Reviewer 2: ```suggestion
The first template we will create will deploy the service for our frontend. Let's create a file called `chart/myapp/templates/frontend-service.yaml` and paste in the following:
```

        app: "backend-selector"
    spec:
      containers:
      - name: backend
        image: "{{ .Values.backend.image.repository }}:{{ .Values.backend.image.tag }}"
        imagePullPolicy: {{ .Values.backend.image.pullPolicy }}
        livenessProbe:
          httpGet:
            path: /health
            port: {{ .Values.backend.service.servicePort }}
          initialDelaySeconds: {{ .Values.backend.livenessProbe.initialDelaySeconds}}
          periodSeconds: {{ .Values.backend.livenessProbe.periodSeconds}}
        ports:
        - containerPort: {{ .Values.backend.service.servicePort}}
        env:
          - name: PORT
            value : "{{ .Values.backend.service.servicePort }}"
          - name: APPLICATION_NAME
            value: "{{ .Release.Name }}"
          - name: MONGO_URL
            value: {{ .Values.backend.services.mongo.url }}
          - name: MONGO_DB_NAME
            value: {{ .Values.backend.services.mongo.name }}
```

Similar to our frontend deployment file, this file creates our backend deployment, with a key difference being the mongoDB information that is passed through to allow for communication with the mongo instance.

### 5. Values file

> For the `chart/myapp/values.yaml` file we are going to split it in 3 sections, have a read of each section and then add them all to your `values.yaml` file

Reviewer 2: ```suggestion
For the `chart/myapp/values.yaml` file we are going to split it into 3 sections, have a read of each section and then add them all to your `values.yaml` file
```
